### PR TITLE
Add `linalg.eigh`, `linalg.eigvalsh` to ChainerX 

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This is work-in-progress pull request!

This PR wraps cuSOLVER's Eigen decomposition.

With reference to #6423 and #6764

### CODE REVIEW

File: chainerx_cc/chainerx/kernels/linalg.h [hunk 1]
```diff
@@ -39,4 +40,11 @@ class SvdKernel : public Kernel {
     virtual void Call(const Array& a, const Array& u, const Array& s, const Array& vt, bool full_matrices) = 0;
 };
 
+class SyevdKernel : public Kernel {
+public:
+    static const char* name() { return "Syevd"; }
+
+    virtual void Call(const Array& a, const Array& w, const Array& v, const std::string& UPLO, bool compute_eigen_vector) = 0;
```
Comment (hvy) [inline chainerx_cc/chainerx/kernels/linalg.h]: How about plural "compute_eigen_vectors"? "compute_v" is an option too.

File: chainerx/_docs/routines.py [hunk 2]
```diff
@@ -664,6 +664,46 @@ def _docs_linalg():
 .. seealso:: :func:`numpy.linalg.pinv`
 """)
 
+    _docs.set_doc(
+        chainerx.linalg.eigh,
+        """eigh(a, UPLO='L')
```
Comment (hvy) [inline chainerx/_docs/routines.py]: Let's add a note to the docs that it's not backproppable, and that if one wants to, should use `eigvalsh`.

**Comment by hvy** [comment]
No hurry if you want to focus on the other PRs first but I left some initial comments.

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 3]
```diff
@@ -325,4 +326,63 @@ Array PseudoInverse(const Array& a, float rcond) {
     return out;
 }
 
+std::tuple<Array, Array> Eigh(const Array& a, const std::string& uplo) {
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: As `uplo` is a binary option, how about simply using a bool, in C++? Although that might compromise readability, it should be faster. We could name it `use_a_lower`, `is_a_lower`, ...?

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 4]
```diff
@@ -325,4 +326,63 @@ Array PseudoInverse(const Array& a, float rcond) {
...
+                const Array& v = bctx.GetRetainedOutput(v_tok);
+
+                const Array& gw = bctx.output_grad(0).has_value() ? *bctx.output_grad(0) : Zeros(w.shape(), a.dtype(), a.device());
+                const Array& gv = bctx.output_grad(1).has_value() ? *bctx.output_grad(1) : Zeros(v.shape(), a.dtype(), a.device());
+
+                Array vt = v.Transpose();
+
+                Array f = ExpandDims(w, 0) - ExpandDims(w, 1);
+                // Invert values of F, and fill the diagonal with 0s.
+                // F has 0s on the diagonal, therefore fill it first with infinity.
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: Let's reduce confusion by using the same "f" instead of "F". If you want to clarify that it's referring to `f`, you could surround it with ticks, as in standard markdown.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 5]
```diff
@@ -409,5 +479,64 @@ public:
...
+        int64_t n = a.shape()[1];
+
+        auto syevd_impl = [&](auto pt) {
+            using T = typename decltype(pt)::type;
+            cuda_internal::DeviceInternals& device_internals = cuda_internal::GetDeviceInternals(static_cast<CudaDevice&>(device));
+
+            auto v_ptr = static_cast<T*>(internal::GetRawOffsetData(v));
+            auto w_ptr = static_cast<T*>(internal::GetRawOffsetData(w));
+
+            cusolverEigMode_t jobz = CUSOLVER_EIG_MODE_NOVECTOR;
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: A major nitpick but how about compacting the code with the ternary operator?

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 6]
```diff
@@ -409,5 +479,64 @@ public:
...
+            if (compute_eigen_vector) {
+                jobz = CUSOLVER_EIG_MODE_VECTOR;
+            }
+
+            // cuSOLVER assumes that arrays are stored in column-major order
+            // The uplo argument is swapped instead of transposing the input matrix
+            cublasFillMode_t uplo = CUBLAS_FILL_MODE_UPPER;
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: Ditto for ternary.

File: chainerx_cc/chainerx/native/native_device/linalg.cc [hunk 7]
```diff
@@ -360,5 +390,78 @@ class NativeSvdKernel : public SvdKernel {
...
+        int64_t m = a.shape()[0];
+        int64_t n = a.shape()[1];
+
+        auto syevd_impl = [&](auto pt) {
+            using T = typename decltype(pt)::type;
+
+            auto v_ptr = static_cast<T*>(internal::GetRawOffsetData(v));
+            auto w_ptr = static_cast<T*>(internal::GetRawOffsetData(w));
+
+            char jobz = 'N';
```
Comment (hvy) [inline chainerx_cc/chainerx/native/native_device/linalg.cc]: Ditto for ternary.

File: chainerx_cc/chainerx/routines/linalg.cc (hunk 3 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/routines/linalg.cc]: It may be faster or may not. However this char/string flag is used in LAPACK and NumPy, I think it is worth keeping C++ interface consistent with those libraries.

File: chainerx_cc/chainerx/kernels/linalg.h (hunk 1 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/kernels/linalg.h]: I changed to be `compute_v`, so it similar to SVD's `compute_uv`.

File: chainerx/_docs/routines.py (hunk 2 above)
Comment (IvanYashchuk) [reply chainerx/_docs/routines.py]: Done. Actually it is the opposite. `eigvalsh` (computes only eigenvalues) is not backproppable because eigenvectors are required.

File: chainerx_cc/chainerx/routines/linalg.cc (hunk 3 above)
Comment (hvy) [reply chainerx_cc/chainerx/routines/linalg.cc]: How about using a char instead of string?

File: chainerx_cc/chainerx/kernels/linalg.h [hunk 8]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <string>
```
Comment (hvy) [inline chainerx_cc/chainerx/kernels/linalg.h]: Let's remove this include now that it's not used.
```suggestion
```

File: chainerx_cc/chainerx/routines/linalg.h [hunk 9]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <string>
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.h]: Ditto.
```suggestion
```

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 10]
```diff
@@ -4,6 +4,7 @@
 #include <cassert>
 #include <cstdint>
 #include <limits>
+#include <string>
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: Ditto.
```suggestion
```

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 11]
```diff
@@ -409,5 +479,58 @@ public:
...
+            cuda_internal::DeviceInternals& device_internals = cuda_internal::GetDeviceInternals(static_cast<CudaDevice&>(device));
+
+            auto v_ptr = static_cast<T*>(internal::GetRawOffsetData(v));
+            auto w_ptr = static_cast<T*>(internal::GetRawOffsetData(w));
+
+            cusolverEigMode_t jobz = compute_v ? CUSOLVER_EIG_MODE_VECTOR : CUSOLVER_EIG_MODE_NOVECTOR;
+
+            // cuSOLVER assumes that arrays are stored in column-major order
+            // The uplo argument is swapped instead of transposing the input matrix
+            cublasFillMode_t uplo_cublas = (uplo == 'U') ? CUBLAS_FILL_MODE_LOWER : CUBLAS_FILL_MODE_UPPER;
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: A nitpick, but for consistency.
```suggestion
            cublasFillMode_t uplo_cublas = uplo == 'U' ? CUBLAS_FILL_MODE_LOWER : CUBLAS_FILL_MODE_UPPER;
```

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu [hunk 12]
```diff
@@ -409,5 +479,58 @@ public:
...
+
+            int buffersize = 0;
+            // When calling Syevd matrix dimensions are swapped instead of transposing the input matrix
+            device_internals.cusolverdn_handle().Call(SyevdBuffersize<T>, jobz, uplo_cublas, n, v_ptr, m, w_ptr, &buffersize);
+
+            Array work = Empty(Shape{buffersize}, dtype, device);
+            auto work_ptr = static_cast<T*>(internal::GetRawOffsetData(work));
+
+            std::shared_ptr<void> devInfo = device.Allocate(sizeof(int));
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: We should probably use `device_info` over `deviceInfo` but I noticed similar code is already merged. We can fix them altogether at a later time, unless you want to send a patch and fix the ongoing ones.

File: chainerx_cc/chainerx/native/native_device/linalg.cc [hunk 13]
```diff
@@ -360,5 +390,70 @@ class NativeSvdKernel : public SvdKernel {
...
+            using T = typename decltype(pt)::type;
+
+            auto v_ptr = static_cast<T*>(internal::GetRawOffsetData(v));
+            auto w_ptr = static_cast<T*>(internal::GetRawOffsetData(w));
+
+            char jobz = compute_v ? 'V' : 'N';
+
+            // LAPACK assumes that arrays are stored in column-major order
+            // The uplo argument is swapped instead of transposing the input matrix
+            char uplo_swapped = (uplo == 'U') ? 'L' : 'U';
```
Comment (hvy) [inline chainerx_cc/chainerx/native/native_device/linalg.cc]: Ditto.
```suggestion
            char uplo_swapped = uplo == 'U' ? 'L' : 'U';
```

File: chainerx_cc/chainerx/native/native_device/linalg.cc [hunk 14]
```diff
@@ -360,5 +390,70 @@ class NativeSvdKernel : public SvdKernel {
...
+
+            int info;
+            int lwork = -1;
+            int liwork = -1;
+            T work_size;
+            int iwork_size;
+
+            // When calling Syevd matrix dimensions are swapped instead of transposing the input matrix
+            Syevd<T>(jobz, uplo_swapped, n, v_ptr, m, w_ptr, &work_size, lwork, &iwork_size, liwork, &info);
+
```
Comment (hvy) [inline chainerx_cc/chainerx/native/native_device/linalg.cc]: Should you check `info` before proceeding to the next LOC?

File: chainerx_cc/chainerx/python/routines.cc [hunk 15]
```diff
@@ -369,6 +369,21 @@ void InitChainerxLinalg(pybind11::module& m) {
             [](const ArrayBodyPtr& a, float rcond) { return MoveArrayBody(PseudoInverse(Array{a}, rcond)); },
             "a"_a,
             "rcond"_a = 1e-15);
+    mlinalg.def(
+            "eigh",
+            [](const ArrayBodyPtr& a, const std::string& UPLO) {
+                std::tuple<Array, Array> wv = Eigh(Array{a}, UPLO.c_str()[0]);
```
Comment (hvy) [inline chainerx_cc/chainerx/python/routines.cc]: We must validate `UPLO` here. Let's in the binding check if it's a string of length 1, and in the routine check that it's actually `'U'` or `'L'`.

File: chainerx_cc/chainerx/python/routines.cc [hunk 16]
```diff
@@ -369,6 +369,21 @@ void InitChainerxLinalg(pybind11::module& m) {
...
+                Array& w = std::get<0>(wv);
+                Array& v = std::get<1>(wv);
+                return std::make_tuple(MoveArrayBody(std::move(w)), MoveArrayBody(std::move(v)));
+            },
+            "a"_a,
+            "UPLO"_a = "L");
+    mlinalg.def(
+            "eigvalsh",
+            [](const ArrayBodyPtr& a, const std::string& UPLO) { return MoveArrayBody(Eigvalsh(Array{a}, UPLO.c_str()[0])); },
```
Comment (hvy) [inline chainerx_cc/chainerx/python/routines.cc]: Ditto regarding `UPLO` validation.

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 17]
```diff
@@ -325,4 +326,63 @@ Array PseudoInverse(const Array& a, float rcond) {
     return out;
 }
 
+std::tuple<Array, Array> Eigh(const Array& a, char uplo) {
+    CheckRankTwoArray(a);
+    CheckSquareMatrix(a);
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: Let's also check `uplo`.

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 18]
```diff
@@ -325,4 +326,63 @@ Array PseudoInverse(const Array& a, float rcond) {
...
+        }
+        bb.Finalize();
+    }
+
+    return std::make_tuple(std::move(w), std::move(v));
+}
+
+Array Eigvalsh(const Array& a, char uplo) {
+    CheckRankTwoArray(a);
+    CheckSquareMatrix(a);
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: Ditto.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 19]
```diff
@@ -351,3 +351,145 @@ def forward_xp(self, inputs, xp):
...
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    chainer.testing.product({
+        'shape': [(), (2, 3), (3, 2)],
+        'in_dtypes': ['float32', 'float64']
+    })
+))
+class TestEigvalshFailing(NumpyLinalgOpTest):
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Let's add tests for invalid `UPLO`, i.e. some other string that "U", "L". We could also test `UPLO=None` if we want to be perfect.

File: chainerx_cc/chainerx/cuda/cuda_device/linalg.cu (hunk 12 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/cuda/cuda_device/linalg.cu]: > unless you want to send a patch and fix the ongoing ones.

I will do this

**Comment by hvy** [comment]
Jenkins, test this please.

File: chainerx_cc/chainerx/native/native_device/linalg.cc (hunk 14 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/native/native_device/linalg.cc]: The first call is a workspace query. The routine only calculates the required sizes of workspace arrays, returns these values, and no error message is issued.
src: https://software.intel.com/en-us/mkl-developer-reference-fortran-syevd

Comment (IvanYashchuk) [reply chainerx_cc/chainerx/native/native_device/linalg.cc]: As written in docs this step is not necessary if we are willing to provide plenty of memory for workspace, but the execution
>though probably not so fast as with a recommended workspace.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py [hunk 20]
```diff
@@ -351,3 +351,155 @@ def forward_xp(self, inputs, xp):
...
+        return a,
+
+    def forward_xp(self, inputs, xp):
+        a, = inputs
+
+        # Input has to be symmetrized for backward test to work
+        def symmetrize(A):
+            L = xp.tril(A)
+            return (L + L.T)/2.
+        a = symmetrize(a)
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Actually, I think gradients should be zero for the unused half. I.e. in the backward definition, if uplo is L, add the upper half to the lower half of the gradient, zeroing out the upper half. And vise versa. Then we do not need this symmetrization.

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 21]
```diff
@@ -325,4 +332,65 @@ Array PseudoInverse(const Array& a, float rcond) {
...
+    CheckUplo(uplo);
+
+    Array w = Empty(Shape{a.shape()[0]}, a.dtype(), a.device());
+    Array v = Empty(a.shape(), a.dtype(), a.device());
+
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<SyevdKernel>(a, w, v, uplo, /*compute_v=*/true);
+        v = v.Transpose();
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: Should we let the kernel be responsible for this transpose?

File: chainerx_cc/chainerx/routines/linalg.cc [hunk 22]
```diff
@@ -325,4 +332,65 @@ Array PseudoInverse(const Array& a, float rcond) {
...
+    }
+
+    // Reference:
+    // Section 3.1 https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+    {
+        BackwardBuilder bb{"eigh", a, {w, v}};
+        if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+            bt.Define([a_tok = bb.RetainInput(0), w_tok = bb.RetainOutput(0), v_tok = bb.RetainOutput(1)](BackwardContext& bctx) {
+                const Array& a = bctx.GetRetainedInput(a_tok);
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/linalg.cc]: Let's not retain `a` just for the dtype and device. We can retain "a_type" only.

**Comment by hvy** [comment]
Sorry for another round of review comments but could you fix the above?

File: chainerx_cc/chainerx/routines/linalg.cc (hunk 21 above)
Comment (IvanYashchuk) [reply chainerx_cc/chainerx/routines/linalg.cc]: I've moved `v = v.Transpose()` to the kernels.
It's done with
`device.backend().CallKernel<CopyKernel>(v.Transpose(), v);`

File: tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py (hunk 20 above)
Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py]: Discussed this with the author outside this thread. I now understand that the current implementation is correct, both from a mathematical as well as a practical (as an autograd framework) standpoint. That this routine only uses either half is an "implementation detail" and should not affect the values of the gradient. It must be noted though that the routine **assumes** that the input indeed is symmetric since we choose not to perform any validation for it.

**Comment by hvy** [comment]
These two routines suffer from the same issue as the one reported here https://github.com/chainer/chainer/pull/7329#issuecomment-522829589, but fails not only for cuda but both native and cuda. NumPy returns empty arrays so we should fix these ones too.

**Comment by IvanYashchuk** [comment]
Inputs with shape (0, 0) are now working. The problem was that the leading dimension argument can't be less than one when calling LAPACK/cuSOLVER.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
We're currently trying to address flaky tests in Chainer (tests that fail quite often, but randomly). I tried to run this tests around 1000 times locally and noticed it is too. Could you possibly take a look at it?

```python
> python -m pytest -rfx -x tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py -k 'Eig or eig'
=================================================== test session starts ====================================================
platform linux -- Python 3.6.1, pytest-4.1.1, py-1.7.0, pluggy-0.7.1
rootdir: /home/vincent/repos/chainer, inifile: setup.cfg
plugins: xdist-1.26.1, repeat-0.7.0, metadata-1.8.0, html-1.20.0, forked-1.0.2
collected 5452 items / 5312 deselected

tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py ..............................F

========================================================= FAILURES =========================================================
________________________ test_Eigh_param_15_{UPLO='L', in_dtypes='float64', shape=(6, 6)}[native:0] ________________________

... 

>           raise AssertionError(f.getvalue())
E           chainer.testing.function_link.FunctionTestError: backward is not implemented correctly
E
E           (caused by)
E           AssertionError: check_backward failed (eps=0.001 atol=1e-05 rtol=0.005)
E           inputs[0]:
E           array([[0.63003572, 0.73949666, 0.30431390, 0.78106292, 0.84906733, 0.13343415],
E                  [0.77523393, 0.72043896, 0.35694555, 0.20228585, 0.1443275 , 0.46263248],
E                  [0.44895464, 0.98560085, 0.45669856, 0.50982009, 0.67137867, 0.71678652],
E                  [0.80079176, 0.49129112, 0.44117445, 0.37909192, 0.24211678, 0.05098325],
E                  [0.85323461, 0.4698262 , 0.24974733, 0.30615825, 0.44239657, 0.95101930],
E                  [0.67536627, 0.92138063, 0.80352809, 0.06498898, 0.07995298, 0.44205577]], shape=(6, 6), dtype=float64, device='native:0')
E           grad_outputs[0]:
E           array([0.69785838, -0.50916007, -0.38296303, -0.21680939, -0.50352705, -0.52525478], shape=(6,), dtype=float64, device='native:0')
E           grad_outputs[1]:
E           array([[-0.42395427, 0.94644767, -0.10463601, 0.45856564, -0.93298551, -0.77470524],
E                  [-0.73019518, 0.37148137, 0.95580271, -0.8257174 , 0.20094679, 0.71898474],
E                  [0.18891407, -0.9121845 , -0.01234777, -0.46408905, 0.96048164, 0.88721807],
E                  [-0.05649843, 0.45397163, 0.71113162, 0.54623588, -0.63286487, -0.15408585],
E                  [-0.95843457, -0.95512276, 0.57650616, 0.54538773, -0.16373386, 0.37241348],
E                  [-0.34651667, 0.99726244, -0.75112784, -0.75493156, -0.3493308 , 0.02286855]], shape=(6, 6), dtype=float64, device='native:0')
E           directions[0]:
E           array([[-0.00618424, 0.08321978, -0.36595368, 0.33599586, -0.10806886, 0.24544865],
E                  [0.11989653, -0.03817258, -0.18779766, 0.13898499, -0.10236622, -0.23798269],
E                  [0.12837853, 0.04892894, -0.27415405, -0.02997159, 0.05766378, -0.05279465],
E                  [0.24968716, -0.07334048, 0.04902351, -0.04082709, 0.25993498, 0.21250433],
E                  [0.10434886, 0.18213748, -0.04959657, 0.08947205, 0.22029317, 0.18337490],
E                  [0.11972817, 0.18409781, -0.20667259, 0.03948545, -0.05358864, 0.13837495]], shape=(6, 6), dtype=float64, device='native:0')
E           gradients (numeric):  array(0.05349447, shape=(), dtype=float64, device='native:0')
E           gradients (backward): array(0.04892034, shape=(), dtype=float64, device='native:0')
E
E           x: numeric gradient, y: backward gradient
E           Not equal to tolerance rtol=0.005, atol=1e-05
E
E           Mismatch: 100%
E           Max absolute difference: 0.00457413
E           Max relative difference: 0.09350161
E            x: array(0.053494)
E            y: array(0.04892)
E
E           assert_allclose failed:
E             shape: () ()
E             dtype: float64 float64
E             i: (0,)
E             x[i]: 0.05349447366484204
E             y[i]: 0.048920342974239195
E             relative error[i]: 0.09350160715372591
E             absolute error[i]: 0.004574130690602848
E             relative tolerance * |y[i]|: 0.000244601714871196
E             absolute tolerance: 1e-05
E             total tolerance: 0.000254601714871196
E           x: 0.05349447
E           y: 0.04892034

chainer/gradient_check.py:537: FunctionTestError
```

**Comment by hvy** [comment]
Thanks for the fix but it's still a bit flaky in my environment (6 failures in 5000 runs). How about relaxing the tolerances?

**Comment by hvy** [comment]
Sorry but the way in which kernels are registered has been changed in https://github.com/chainer/chainer/pull/7970. Could you make similar changes to this PR?

**Comment by hvy** [comment]
Modified the input construction using https://github.com/chainer/chainer/pull/8077 and it seems to address the flakiness. Could you adopt it?

**Comment by hvy** [comment]
This should be stable for the backward tests now. LGTM!

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by IvanYashchuk** [comment]
Not related
```
=========================== short test summary info ============================
FAIL ../../../repo/tests/chainerx_tests/unit_tests/routines_tests/test_creation.py::test_Meshgrid_param_5_{input_arrs=6}_param_1_{indexing='ij'}[float16-cuda:0]
===== 1 failed, 126209 passed, 5672 skipped, 41 xfailed in 2009.64 seconds =====
```

**Comment by hvy** [comment]
Jenkins, test this please.
